pkg/sql/sem/tree/expr.go

```diff
@@ -1415,6 +1442,59 @@ const (
 	CastPrepend
 )
 
+// UnresolvedCastExpr represents a CAST(expr AS type) expression with a
+// possibly unresolved type reference.
+type UnresolvedCastExpr struct {
```

Comment:
do you know why we can't just make `CastExpr` have ResolvableTypeReference?

Reply:
There's no technical reason, but it involves changing alot more code -- alot of places make cast exprs, then we have to plumb type resolvers to all those places. Ideally once we typecheck there's no need to resolve anymore types, so that's what I went with here. It should only be before typechecking that if someone has a *types.T they need to wrap it with KnownType before doing anything.

Reply:
What about this: We can rename `CastExpr.Type` to `CastExpr.UnresolvedType` (and type it as `ResolvableTypeReference`), and then switch all post-resolution users to call `ResolvedType()` instead. This lets us get rid of `UnresolvedCastExpr` without much disturbance.